snapcraft/storeapi/__init__.py

```diff
@@ -72,3 +133,89 @@ def login(self, email, password, token_name, one_time_password=None):
     def logout(self):
         self.conf.clear()
         self.conf.save()
+
+    def download(self, snap_name, channel, download_path, arch=None):
+        if arch is None:
+            arch = snapcraft.ProjectOptions().deb_arch
+
+        package = self.cpi.search_package(snap_name, channel, arch)
+        if package is None:
+            raise SnapNotFoundError(snap_name, channel, arch)
+        self._download_snap(
+            snap_name, channel, arch, download_path,
+            package['download_url'], package['download_sha512'])
+
+    def _download_snap(self, name, channel, arch, download_path,
+                       download_url, expected_sha512):
+        if self._is_downloaded(download_path, expected_sha512):
+            logger.info('Already downloaded {} at {}'.format(
+                name, download_path))
+            return
+        logger.info('Downloading {}'.format(name, download_path))
+        download = self.cpi.get(download_url)
+        with open(download_path, 'wb') as f:
+            # FIXME: Cough, we may want to buffer here (and a progress bar
+            # would be nice) -- vila 2016-04-26
+            f.write(download.content)
+        if self._is_downloaded(download_path, expected_sha512):
+            logger.info('Successfully downloaded {} at {}'.format(
+                name, download_path))
+        else:
+            raise SHAMismatchError(download_path, expected_sha512)
+
+    def _is_downloaded(self, path, expected_sha512):
+        if not os.path.exists(path):
+            return False
+
+        file_sum = hashlib.sha512()
+        with open(path, 'rb') as f:
+            for file_chunk in iter(
+                    lambda: f.read(file_sum.block_size * 128), b''):
+                file_sum.update(file_chunk)
+        return expected_sha512 == file_sum.hexdigest()
+
+
+class CPIClient(Client):
```

Comment:
Not fond of this class name given that the name CPI is sort of something we don't want to expose outside the store